### PR TITLE
rpc graceful shutdown

### PR DESCRIPTION
**Pull request by beef9999**
rpc支持graceful shutdown，即等待客户端主动关闭链接。
配合socket server现有的terminate（即只关listen fd，阻止建立新链接；不关active connection），可以用如下代码实现优雅关闭：

```c++
socket_server->terminate();
delete socket_server;
rpc_skeletion->shutdown(true);
delete rpc_skeletion;
```

### CODE REVIEW

File: rpc/rpc.cpp [hunk 1]
```diff
@@ -332,24 +330,23 @@ namespace rpc {
...
             while(m_running)
             {
                 Context context(this, stream);
                 context.stream_serv_count = &stream_serv_count;
                 context.stream_cv = &stream_cv;
-                context.w_lock = w_lock;
+                context.w_lock = &w_lock;
                 node.thread = CURRENT;
                 int ret = context.read_request();
-                ERRNO err;
```
Comment (beef9999) [inline rpc/rpc.cpp:352]: 没有用到

File: rpc/rpc.cpp [hunk 2]
```diff
@@ -381,48 +373,55 @@ namespace rpc {
             auto ctx = (Context*)args_;
             Context context(std::move(*ctx));
             ctx->got_it = true;
-            thread_yield_to(nullptr);
+            thread_yield();
```
Comment (beef9999) [inline rpc/rpc.cpp:376]: thread_yield_to 第一步会有一个unlikely的判断，thread_yield 是直接调用

File: rpc/rpc.cpp [hunk 3]
```diff
@@ -381,48 +373,55 @@ namespace rpc {
...
-                if (x->thread)
+            if (method != ShutDownMethod::Graceful) {
+                m_running = false;
+            }
+            std::vector<thread*> stream_threads;
+            for (const auto& x: m_list) {
+                if (!x->thread)
+                    continue;
+                if (method == ShutDownMethod::Graceful) {
+                    stream_threads.push_back(x->thread);
```
Comment (beef9999) [inline rpc/rpc.cpp:393]: 不能在m_list一边变化时一边join，因此先保存一下

File: rpc/rpc.cpp [hunk 4]
```diff
@@ -222,7 +222,7 @@ namespace rpc {
             bool got_it;
             int* stream_serv_count;
             photon::condition_variable *stream_cv;
-            std::shared_ptr<photon::mutex> w_lock;
+            photon::mutex* w_lock;
```
Comment (beef9999) [inline rpc/rpc.cpp:226]: 现在的代码中已经有cv，能够保证mutex在所有request完成后再析构，所以这里不需要用共享指针了。

File: rpc/rpc.cpp [hunk 5]
```diff
@@ -314,43 +313,42 @@ namespace rpc {
...
-            if (!m_running)
-                LOG_ERROR_RETURN(ENOTSUP, -1, "the skeleton has closed");
-
             ThreadLink node;
             m_list.push_back(&node);
             DEFER(m_list.erase(&node));
-            DEFER(if (ownership) delete stream;);
+            DEFER(if (ownership) {
+                delete stream;
+                stream = nullptr;
```
Comment (beef9999) [inline rpc/rpc.cpp:329]: 防止被socket server double delete

File: rpc/rpc.cpp [hunk 6]
```diff
@@ -381,48 +374,47 @@ namespace rpc {
...
+                thread_enable_join(x->thread);
+                if (!passive) {
                     thread_interrupt(x->thread);
-            // it should confirm that all threads are finished
-            // or m_list may not destruct correctly
-            while (m_serving_count > 0) {
-                // means shutdown called by rpc serve, should return to give chance to shutdown
-                if ((m_serving_count == 1) && (m_list.front()->thread == nullptr))
-                    return 0;
-                m_cond_served.wait_no_lock();
```
Comment (beef9999) [inline rpc/rpc.cpp:407]: 现有代码已经可以保证在stream析构前等待所有requests结束，因此这段可以删了

File: rpc/rpc.cpp (hunk 5 above)
Comment (lihuiba) [reply rpc/rpc.cpp:330]: As ```stream``` is a local variable, assigning it with nullptr won't affect the caller side.

File: rpc/rpc.cpp [hunk 7]
```diff
@@ -381,48 +375,47 @@ namespace rpc {
...
-        virtual int shutdown() override
+        virtual int shutdown(bool passive) override
         {
-            m_running = false;
-            for (const auto& x: m_list)
-                if (x->thread)
+            if (!passive) {
+                m_running = false;
+            }
+            std::vector<thread*> stream_threads;
```
Comment (lihuiba) [inline rpc/rpc.cpp:390]: We can avoid copying the threads out by:
```cpp
while(m_list) {
    // wait for the 1st thread in m_list to complete
}
```
if we need to wait for them.

File: rpc/rpc.h [hunk 8]
```diff
@@ -168,9 +168,16 @@ namespace rpc
...
         virtual void set_allocator(IOAlloc allocation) = 0;
 
-        virtual int shutdown_no_wait() = 0;
+        /**
+         * @brief Shutdown the rpc server from outside.
+         * @param passive Will wait until all clients close the connections
+         * @warning DO NOT invoke this function within the RPC request.
+         *          You should create a thread to invoke it, or just use shutdown_inplace.
+         */
+        virtual int shutdown(bool passive = false) = 0;
```
Comment (lihuiba) [inline rpc/rpc.h:177]:
```cpp
bool no_more_requests = true, bool wait = true
```
in order to behave as before, and allow for more control

File: rpc/rpc.h [hunk 9]
```diff
@@ -168,9 +168,16 @@ namespace rpc
...
 
-        virtual int shutdown() = 0;
+        /** @brief Shutdown the server within a RPC request */
+        virtual void shutdown_inplace(bool passive = false) = 0;
```
Comment (lihuiba) [inline rpc/rpc.h:180]: shutdown_inplace() seems not necessary, because all handlers are async by themselves now.

File: rpc/rpc.h [hunk 10]
```diff
@@ -168,9 +168,16 @@ namespace rpc
         // the default allocator is defined in iovector.h/cpp
         virtual void set_allocator(IOAlloc allocation) = 0;
 
-        virtual int shutdown_no_wait() = 0;
```
Comment (lihuiba) [inline rpc/rpc.h:171]:
```cpp
int shutdown_no_wait() {
    return shutdown(true, false);
}
```

File: rpc/rpc.h [hunk 11]
```diff
@@ -161,16 +161,25 @@ namespace rpc
         virtual int set_close_notify(Notifier notifier) = 0;
 
         // can be invoked concurrently by multiple threads
-        // if have the ownership of stream, `serve` will delete it before exit
-        virtual int serve(IStream* stream, bool ownership_stream = false) = 0;
+        virtual int serve(IStream* stream) = 0;
+
+        [[deprecated]]
```
Comment (lihuiba) [inline rpc/rpc.h:166]: use ```__attribute__((deprecated))``` to fit for C++11 (and gcc 4.9)

similar for ```new_skeleton()```

File: rpc/rpc.cpp [hunk 12]
```diff
@@ -381,48 +370,41 @@ namespace rpc {
...
-            // it should confirm that all threads are finished
-            // or m_list may not destruct correctly
-            while (m_serving_count > 0) {
-                // means shutdown called by rpc serve, should return to give chance to shutdown
-                if ((m_serving_count == 1) && (m_list.front()->thread == nullptr))
-                    return 0;
-                m_cond_served.wait_no_lock();
+        virtual int shutdown(bool no_more_requests) override {
+            if (no_more_requests)
+                m_running = false;
```
Comment (lihuiba) [inline rpc/rpc.cpp:382]: m_running = !no_more_requests;

File: rpc/rpc.cpp [hunk 13]
```diff
@@ -381,48 +370,41 @@ namespace rpc {
...
+                }
+                // Wait all streams destructed. Their attached RPC requests are finished as well.
+                thread_join((join_handle*) m_list.front()->thread);
             }
-            while (!m_list.empty())
-                thread_usleep(1000);
             return 0;
         }
+        void shutdown_no_wait(bool no_more_requests) override {
+            thread_create11(&SkeletonImpl::shutdown, this, no_more_requests);
```
Comment (lihuiba) [inline rpc/rpc.cpp:394]: simply ```m_running = ! no_more_requests;```

File: rpc/rpc.cpp [hunk 14]
```diff
@@ -381,48 +370,41 @@ namespace rpc {
...
-            while (m_serving_count > 0) {
-                // means shutdown called by rpc serve, should return to give chance to shutdown
-                if ((m_serving_count == 1) && (m_list.front()->thread == nullptr))
-                    return 0;
-                m_cond_served.wait_no_lock();
+        virtual int shutdown(bool no_more_requests) override {
+            if (no_more_requests)
+                m_running = false;
+            while (m_list) {
+                thread_enable_join(m_list.front()->thread);
```
Comment (lihuiba) [inline rpc/rpc.cpp:384]: ```auto th = m_list.front()->thread;```
then use ```th``` instead.

File: rpc/rpc.h (hunk 11 above)
Comment (beef9999) [reply rpc/rpc.h]: It's contradictory to support Clang/Windows and old gcc compiler at the same time

**Review by lihuiba — APPROVED** [review]

**Pull request merged**